File: clients/java_graal/src/main/java/ai_cup_22/model/Item.java
package ai_cup_22.model;

import ai_cup_22.util.StreamUtil;

/**
 * Lootable item
 */
public abstract class Item {
    /**
     * Write Item to output stream
     */
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;

    /**
     * Read Item from input stream
     */
    public static Item readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case Weapon.TAG:
                return Weapon.readFrom(stream);
            case ShieldPotions.TAG:
                return ShieldPotions.readFrom(stream);
            case Ammo.TAG:
                return Ammo.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected tag value");
        }
    }

    /**
     * Weapon
     */
    public static class Weapon extends Item {
        public static final int TAG = 0;
    
        /**
         * Weapon type index (starting with 0)
         */
        private int typeIndex;
    
        /**
         * Weapon type index (starting with 0)
         */
        public int getTypeIndex() {
            return typeIndex;
        }
    
        /**
         * Weapon type index (starting with 0)
         */
        public void setTypeIndex(int value) {
            this.typeIndex = value;
        }
    
        public Weapon(int typeIndex) {
            this.typeIndex = typeIndex;
        }
    
        /**
         * Read Weapon from input stream
         */
        public static Weapon readFrom(java.io.InputStream stream) throws java.io.IOException {
            int typeIndex;
            typeIndex = StreamUtil.readInt(stream);
            return new Weapon(typeIndex);
        }
    
        /**
         * Write Weapon to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, typeIndex);
        }
    
        /**
         * Get string representation of Weapon
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("Weapon { ");
            stringBuilder.append("typeIndex: ");
            stringBuilder.append(String.valueOf(typeIndex));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Shield potions
     */
    public static class ShieldPotions extends Item {
        public static final int TAG = 1;
    
        /**
         * Amount of potions
         */
        private int amount;
    
        /**
         * Amount of potions
         */
        public int getAmount() {
            return amount;
        }
    
        /**
         * Amount of potions
         */
        public void setAmount(int value) {
            this.amount = value;
        }
    
        public ShieldPotions(int amount) {
            this.amount = amount;
        }
    
        /**
         * Read ShieldPotions from input stream
         */
        public static ShieldPotions readFrom(java.io.InputStream stream) throws java.io.IOException {
            int amount;
            amount = StreamUtil.readInt(stream);
            return new ShieldPotions(amount);
        }
    
        /**
         * Write ShieldPotions to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, amount);
        }
    
        /**
         * Get string representation of ShieldPotions
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("ShieldPotions { ");
            stringBuilder.append("amount: ");
            stringBuilder.append(String.valueOf(amount));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }

    /**
     * Ammo
     */
    public static class Ammo extends Item {
        public static final int TAG = 2;
    
        /**
         * Weapon type index (starting with 0)
         */
        private int weaponTypeIndex;
    
        /**
         * Weapon type index (starting with 0)
         */
        public int getWeaponTypeIndex() {
            return weaponTypeIndex;
        }
    
        /**
         * Weapon type index (starting with 0)
         */
        public void setWeaponTypeIndex(int value) {
            this.weaponTypeIndex = value;
        }
        /**
         * Amount of ammo
         */
        private int amount;
    
        /**
         * Amount of ammo
         */
        public int getAmount() {
            return amount;
        }
    
        /**
         * Amount of ammo
         */
        public void setAmount(int value) {
            this.amount = value;
        }
    
        public Ammo(int weaponTypeIndex, int amount) {
            this.weaponTypeIndex = weaponTypeIndex;
            this.amount = amount;
        }
    
        /**
         * Read Ammo from input stream
         */
        public static Ammo readFrom(java.io.InputStream stream) throws java.io.IOException {
            int weaponTypeIndex;
            weaponTypeIndex = StreamUtil.readInt(stream);
            int amount;
            amount = StreamUtil.readInt(stream);
            return new Ammo(weaponTypeIndex, amount);
        }
    
        /**
         * Write Ammo to output stream
         */
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, weaponTypeIndex);
            StreamUtil.writeInt(stream, amount);
        }
    
        /**
         * Get string representation of Ammo
         */
        @Override
        public String toString() {
            StringBuilder stringBuilder = new StringBuilder("Ammo { ");
            stringBuilder.append("weaponTypeIndex: ");
            stringBuilder.append(String.valueOf(weaponTypeIndex));
            stringBuilder.append(", ");
            stringBuilder.append("amount: ");
            stringBuilder.append(String.valueOf(amount));
            stringBuilder.append(" }");
            return stringBuilder.toString();
        }
    }
}

File: clients/kotlin_graal/src/main/kotlin/ai_cup_22/model/Item.kt
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Lootable item
 */
abstract class Item {
    /**
     * Write Item to output stream
     */
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)

    companion object {
        /**
         * Read Item from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Item {
            when (StreamUtil.readInt(stream)) {
                Weapon.TAG -> return Weapon.readFrom(stream)
                ShieldPotions.TAG -> return ShieldPotions.readFrom(stream)
                Ammo.TAG -> return Ammo.readFrom(stream)
                else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }

    /**
     * Weapon
     */
    class Weapon : Item {
        /**
         * Weapon type index (starting with 0)
         */
        var typeIndex: Int
    
        constructor(typeIndex: Int) {
            this.typeIndex = typeIndex
        }
    
        /**
         * Write Weapon to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, typeIndex)
        }
    
        /**
         * Get string representation of Weapon
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("Weapon { ")
            stringBuilder.append("typeIndex: ")
            stringBuilder.append(typeIndex)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 0
    
            /**
             * Read Weapon from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): Weapon {
                var typeIndex: Int
                typeIndex = StreamUtil.readInt(stream)
                return Weapon(typeIndex)
            }
        }
    }

    /**
     * Shield potions
     */
    class ShieldPotions : Item {
        /**
         * Amount of potions
         */
        var amount: Int
    
        constructor(amount: Int) {
            this.amount = amount
        }
    
        /**
         * Write ShieldPotions to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, amount)
        }
    
        /**
         * Get string representation of ShieldPotions
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("ShieldPotions { ")
            stringBuilder.append("amount: ")
            stringBuilder.append(amount)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 1
    
            /**
             * Read ShieldPotions from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): ShieldPotions {
                var amount: Int
                amount = StreamUtil.readInt(stream)
                return ShieldPotions(amount)
            }
        }
    }

    /**
     * Ammo
     */
    class Ammo : Item {
        /**
         * Weapon type index (starting with 0)
         */
        var weaponTypeIndex: Int
        /**
         * Amount of ammo
         */
        var amount: Int
    
        constructor(weaponTypeIndex: Int, amount: Int) {
            this.weaponTypeIndex = weaponTypeIndex
            this.amount = amount
        }
    
        /**
         * Write Ammo to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, weaponTypeIndex)
            StreamUtil.writeInt(stream, amount)
        }
    
        /**
         * Get string representation of Ammo
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("Ammo { ")
            stringBuilder.append("weaponTypeIndex: ")
            stringBuilder.append(weaponTypeIndex)
            stringBuilder.append(", ")
            stringBuilder.append("amount: ")
            stringBuilder.append(amount)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 2
    
            /**
             * Read Ammo from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): Ammo {
                var weaponTypeIndex: Int
                weaponTypeIndex = StreamUtil.readInt(stream)
                var amount: Int
                amount = StreamUtil.readInt(stream)
                return Ammo(weaponTypeIndex, amount)
            }
        }
    }
}